From 4764471c3d1629fe9fa81133be8a39ef0911760a Mon Sep 17 00:00:00 2001
From: Robert Lehmann <robert.lehmann@sitec-systems.de>
Date: Wed, 3 Feb 2016 08:33:54 +0100
Subject: [PATCH 13/14] S2: Add reboot over KILL_MAIN_PWR GPIO

Now an reboot will be triggered by pushing the KILL_MAIN_PWR GPIO to
high. The hole board (not only the processor) will be restarted.

Ticket #924
---
 arch/arm/mach-imx/system.c | 16 ++++++++++++++++
 1 file changed, 16 insertions(+)

diff --git a/arch/arm/mach-imx/system.c b/arch/arm/mach-imx/system.c
index 2e3d6a8..bfcebba 100644
--- a/arch/arm/mach-imx/system.c
+++ b/arch/arm/mach-imx/system.c
@@ -23,6 +23,7 @@
 #include <linux/delay.h>
 #include <linux/of.h>
 #include <linux/of_address.h>
+#include <linux/gpio.h>
 
 #include <asm/system_misc.h>
 #include <asm/proc-fns.h>
@@ -32,6 +33,8 @@
 #include "common.h"
 #include "hardware.h"
 
+#define KILL_MAIN_PWR IMX_GPIO_NR(3,11)
+
 static void __iomem *wdog_base;
 static struct clk *wdog_clk;
 static u32 wdog_source = 1; /* use WDOG1 default */
@@ -74,6 +77,19 @@ void mxc_restart(enum reboot_mode mode, const char *cmd)
 	__raw_writew(wcr_enable, wdog_base);
 	__raw_writew(wcr_enable, wdog_base);
 
+    /*
+     * Reset the S2 with an High on the GPIO KILL_MAIN_PWR. This remains the main
+     * power swichter to shutdown. After a while (under 1 second) the main switcher
+     * comes up again and the board will reboot
+     */
+    if (of_machine_is_compatible("fsl,imx6sx-s2")) {
+        if (gpio_is_valid(KILL_MAIN_PWR) &&
+            !gpio_request_one(KILL_MAIN_PWR, GPIOF_OUT_INIT_HIGH, "kill_main_pwr"))
+            pr_info("GPIO for Killing Main power exported\n");
+        else
+            pr_warn("Failed to register GPIO %i for killing main power\n", KILL_MAIN_PWR);
+    }
+
 	/* wait for reset to assert... */
 	mdelay(500);
 
-- 
1.9.1

